File: backend/app/schemas/crm.py
"""CRM (leads, messages, templates) schemas."""
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.lead import LEAD_STAGES


class LeadCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=40)
    source: str = "manual"
    trip_id: uuid.UUID | None = None
    notes: str | None = None


class LeadOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    email: str | None
    phone: str | None
    source: str
    stage: str
    score: int
    trip_id: uuid.UUID | None
    notes: str | None
    last_contacted_at: datetime | None
    created_at: datetime


class MessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    channel: str
    direction: str
    subject: str | None
    body: str
    status: str
    created_at: datetime


class ActivityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    type: str
    content: str | None
    meta: dict[str, Any]
    created_at: datetime


class LeadDetail(LeadOut):
    messages: list[MessageOut] = []
    activities: list[ActivityOut] = []


class StageUpdate(BaseModel):
    stage: str = Field(pattern="^(" + "|".join(LEAD_STAGES) + ")$")


class ReplyIn(BaseModel):
    body: str = Field(min_length=1)
    channel: str = "email"


class InboundIn(BaseModel):
    body: str = Field(min_length=1)
    channel: str = "email"


class CaptureIn(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    email: EmailStr | None = None
    phone: str | None = None
    trip_id: uuid.UUID | None = None
    message: str | None = None


class TemplateCreate(BaseModel):
    name: str
    channel: str = "email"
    subject: str | None = None
    body: str
    step: int = 0


class TemplateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    channel: str
    subject: str | None
    body: str
    step: int

File: frontend/src/lib/crm.ts
/** Types + REST calls for the CRM (leads, messages, pipeline). */
import { api } from "./api";

export const STAGES = [
  "new",
  "contacted",
  "interested",
  "proposal",
  "negotiation",
  "won",
  "lost",
] as const;
export type Stage = (typeof STAGES)[number];

export const STAGE_LABEL: Record<Stage, string> = {
  new: "New",
  contacted: "Contacted",
  interested: "Interested",
  proposal: "Proposal",
  negotiation: "Negotiation",
  won: "Won",
  lost: "Lost",
};

export interface Lead {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  source: string;
  stage: Stage;
  score: number;
  trip_id: string | null;
  notes: string | null;
  last_contacted_at: string | null;
  created_at: string;
}

export interface LeadMessage {
  id: string;
  channel: string;
  direction: "outbound" | "inbound";
  subject: string | null;
  body: string;
  status: string;
  created_at: string;
}

export interface LeadActivity {
  id: string;
  type: string;
  content: string | null;
  meta: Record<string, unknown>;
  created_at: string;
}

export interface LeadDetail extends Lead {
  messages: LeadMessage[];
  activities: LeadActivity[];
}

export const listLeads = () => api<Lead[]>("/api/v1/leads");
export const createLead = (data: { name: string; email?: string; phone?: string }) =>
  api<Lead>("/api/v1/leads", { method: "POST", body: JSON.stringify(data) });
export const getLead = (id: string) => api<LeadDetail>(`/api/v1/leads/${id}`);
export const updateStage = (id: string, stage: Stage) =>
  api<Lead>(`/api/v1/leads/${id}/stage`, {
    method: "PATCH",
    body: JSON.stringify({ stage }),
  });
export const replyToLead = (id: string, body: string, channel = "email") =>
  api<LeadMessage>(`/api/v1/leads/${id}/reply`, {
    method: "POST",
    body: JSON.stringify({ body, channel }),
  });
export const simulateInbound = (id: string, body: string, channel = "email") =>
  api<Lead>(`/api/v1/leads/${id}/inbound`, {
    method: "POST",
    body: JSON.stringify({ body, channel }),
  });
export const setAutoFollowup = (enabled: boolean) =>
  api<{ auto_followup: boolean }>(
    `/api/v1/settings/auto-followup?enabled=${enabled}`,
    { method: "POST" },
  );
